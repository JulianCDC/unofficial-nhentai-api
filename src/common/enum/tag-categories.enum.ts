export enum CATEGORIES {
  tags,
  artists,
  characters,
  parodies,
  groups,
}
