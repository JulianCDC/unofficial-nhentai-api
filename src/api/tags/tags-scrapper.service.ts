import { Injectable, Logger } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Cron, CronExpression } from '@nestjs/schedule';
import { Model } from 'mongoose';
import { CATEGORIES } from 'src/shared/enum/tag-categories.enum';
import { HtmlTag } from 'src/shared/types/html-tag.entity';
import { HtmlParserService } from '../../html-parser/html-parser/html-parser.service';
import { getPages } from '../../shared/helpers/get-pagination.helper';
import { parseTags } from '../../shared/helpers/parse-tags.helper';
import { TagDocument } from '../../shared/schemas/tag.schema';

@Injectable()
export class TagsScrapperService {
  private readonly categories = Object.values(CATEGORIES);
  private readonly logger = new Logger(TagsScrapperService.name);

  constructor(
    @InjectModel(TagDocument.name) private tagModel: Model<TagDocument>,
    private htmlParser: HtmlParserService,
  ) {}

  @Cron(CronExpression.EVERY_WEEK)
  parseAllTags(): void {
    this.categories.forEach(async category => {
      const tags: HtmlTag[] = [];
      const $ = await this.htmlParser.parse(`https://nhentai.net/${category}`);

      const numberOfPages = getPages($, 0);

      for (let i = 1; i <= numberOfPages; i++) {
        const uri = `https://nhentai.net/${category}/?page=${i}`;
        const $page = await this.htmlParser.parse(uri);

        tags.push(...parseTags($page('.tag'), $page));
      }

      tags.forEach(tag => {
        this.tagModel.findOneAndUpdate(
          { name: tag.name },
          { ...tag, category: category },
          {
            upsert: true,
          },
          err => {
            this.logger.error(err);
          },
        );
      });
    });
  }
}
