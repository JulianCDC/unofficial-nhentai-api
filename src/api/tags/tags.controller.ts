import {
  CacheInterceptor,
  Controller,
  Get,
  NotFoundException,
  Param,
  Query,
  UseInterceptors,
} from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiNotFoundResponse,
  ApiOperation,
  ApiTags,
} from '@nestjs/swagger';
import { Tag } from 'src/shared/types/tag.entity';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagsService } from './tags.service';
import { GetCategoryPageByLetterDto } from './types/get-category-page-by-letter.dto';
import { GetCategoryPageDto } from './types/get-category-page.dto';
import { GetCategoryDto } from './types/get-category.dto';
import { GetTagByIdDto } from './types/get-tag-by-id.dto';
import { SearchCategoryDto } from './types/search-category.dto';
import { TagListEntity } from './types/tag-list.entity';

@ApiTags('tags')
@Controller('tags')
@UseInterceptors(CacheInterceptor)
export class TagsController {
  constructor(private tagsService: TagsService) {}

  @ApiOperation({
    summary: 'get tag infos for requested id',
  })
  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'no tag found with this id' })
  @Get(':id/details')
  async getTagById(@Param() params: GetTagByIdDto): Promise<Tag> {
    return this.tagsService.getTagById(params.id);
  }

  @ApiOperation({
    summary: 'search for a tag by name and optionaly by category',
    description:
      'returns a list sorted depending on the position of the search query in the tag name',
  })
  @ApiBadRequestResponse()
  @Get('search')
  async searchTag(@Query() query: SearchCategoryDto): Promise<Tag[]> {
    return this.tagsService.search(query.q, query.page, query.category);
  }

  @ApiOperation({
    summary: 'get possible values for category',
  })
  @Get('categories')
  getCategories(): string[] {
    return Object.values(CATEGORIES);
  }

  @ApiOperation({
    summary: 'get all existing tags in specified category',
  })
  @ApiBadRequestResponse()
  @ApiNotFoundResponse({ description: 'requested page does not exist' })
  @Get('list/:category')
  async getTagsByCategory(
    @Param() params: GetCategoryDto,
    @Query() query: GetCategoryPageDto,
  ): Promise<TagListEntity> {
    let tags: Tag[];
    const numberOfPages = await this.tagsService.getPageCount(params.category);

    if (query.page > numberOfPages) {
      throw new NotFoundException();
    }

    if (query.popular) {
      tags = await this.tagsService.getTagsByPopularity(
        params.category,
        query.page,
      );
    } else {
      tags = await this.tagsService.getTags(params.category, query.page);
    }

    return {
      data: tags,
      pagination: {
        page: query.page,
        total: numberOfPages,
      },
    };
  }

  @ApiOperation({
    summary:
      'get all existing tags starting by specified letter in specified category',
  })
  @ApiBadRequestResponse()
  @Get('list/:category/:letter')
  async getTagsByCategoryAndLetter(
    @Param() params: GetCategoryPageByLetterDto,
  ): Promise<Tag[]> {
    return this.tagsService.getTagsByLetter(params.category, params.letter);
  }
}
