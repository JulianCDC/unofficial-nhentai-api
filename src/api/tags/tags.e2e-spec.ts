jest.mock('./tags.service');

import { INestApplication } from '@nestjs/common';
import { getModelToken } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import { random } from 'faker';
import { curry } from 'lodash';
import * as request from 'supertest';
import { createTestingApp, testRequests } from '../../../test/helpers/e2e';
import { CATEGORIES } from '../../shared/enum/tag-categories.enum';
import { TagDocument } from '../../shared/schemas/tag.schema';
import { TagsModule } from './tags.module';
import { TagsService } from './tags.service';
import {
  mockedTags,
  TagsService as MockedTagsService,
} from './__mocks__/tags.service';

describe('TagsController (e2e)', () => {
  let app: INestApplication;
  let tagsService: TagsService;

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [TagsModule],
    })
      .overrideProvider(TagsService)
      .useClass(MockedTagsService)
      .overrideProvider(getModelToken(TagDocument.name))
      .useValue({})
      .compile();

    tagsService = moduleRef.get(TagsService);

    app = createTestingApp(moduleRef);
    await app.init();
  });

  describe('/tags/list/:category (GET)', () => {
    it('sort by popularity', async () => {
      const result = await tagsService.getTagsByPopularity(
        CATEGORIES.artists,
        1,
      );

      return request(app.getHttpServer())
        .get('/tags/list/artists?popular=true&page=1')
        .expect(200)
        .expect({
          data: result,
          pagination: {
            page: 1,
            total: await tagsService.getPageCount(CATEGORIES.artists),
          },
        });
    });

    it('sort by name', async () => {
      const result = await tagsService.getTags(CATEGORIES.artists, 1);

      return request(app.getHttpServer())
        .get('/tags/list/artists?popular=false&page=1')
        .expect(200)
        .expect({
          data: result,
          pagination: {
            page: 1,
            total: await tagsService.getPageCount(CATEGORIES.artists),
          },
        });
    });

    it('throws 404 error when requested page does not exist', async () => {
      const numberOfPages = await tagsService.getPageCount(CATEGORIES.artists);

      return request(app.getHttpServer())
        .get(`/tags/list/artists?popular=false&page=${numberOfPages + 1}`)
        .expect(404);
    });

    it('throws 400 error on wrong parameters values', async () => {
      const httpServer = app.getHttpServer();
      const badRequests: string[] = [
        '/tags/list/notacategory?popular=false&page=1',
        '/tags/list/artists?popular=false&page=0',
        '/tags/list/artists?popular=false&page=1.50',
        '/tags/list/artists?popular=false&page=notanumber',
        '/tags/list/artists?popular=false&page=-1',
      ];

      await testRequests(httpServer, badRequests, 400);
    });

    it('converts parameters types', async () => {
      const httpServer = app.getHttpServer();
      const byPopularity = curry(tagsService.getTagsByPopularity)(
        CATEGORIES.artists,
      );
      const byName = curry(tagsService.getTags)(CATEGORIES.artists);

      const shouldOrderByPopularity: string[] = [
        '/tags/list/artists?popular=true&page=1',
        '/tags/list/artists?popular=1&page=1',
        '/tags/list/artists?popular=anything&page=1',
      ];
      const shouldOrderByName: string[] = [
        '/tags/list/artists?popular=false&page=1',
        '/tags/list/artists?page=1',
      ];
      const shouldWorks: string[] = [
        '/tags/list/artists',
        '/tags/list/artists?page=1',
      ];

      for (const req of shouldOrderByPopularity) {
        await request(httpServer)
          .get(req)
          .expect(200)
          .then(async res => {
            expect(res.body.data).toStrictEqual(await byPopularity(1));
          });
      }

      for (const req of shouldOrderByName) {
        await request(httpServer)
          .get(req)
          .expect(200)
          .then(async res => {
            expect(res.body.data).toStrictEqual(await byName(1));
          });
      }

      for (const req of shouldWorks) {
        await request(httpServer)
          .get(req)
          .expect(200);
      }
    });
  });

  describe('/tags/list/:category/:letter (GET)', () => {
    it('returns all tags starting with specified letter, sorted by name', async () => {
      const result = await tagsService.getTagsByLetter(CATEGORIES.artists, 'A');

      return request(app.getHttpServer())
        .get('/tags/list/artists/a')
        .expect(200)
        .expect(result);
    });

    it('is case insensitive', async () => {
      const result = await tagsService.getTagsByLetter(CATEGORIES.artists, 'A');

      await request(app.getHttpServer())
        .get('/tags/list/artists/a')
        .expect(200)
        .expect(result);

      await request(app.getHttpServer())
        .get('/tags/list/artists/A')
        .expect(200)
        .expect(result);
    });

    it('throws 400 error on wrong parameters values', async () => {
      const httpServer = app.getHttpServer();
      const badRequests: string[] = [
        '/tags/list/notacategory/a',
        '/tags/list/artists/notaletter',
      ];

      await testRequests(httpServer, badRequests, 400);
    });
  });

  describe('/tags/details/:id (GET)', () => {
    it('returns requested tag', async () => {
      const result = await tagsService.getTagById(
        random.arrayElement(mockedTags).id,
      );

      await request(app.getHttpServer())
        .get(`/tags/details/${result.id}`)
        .expect(200)
        .expect(result);
    });

    it('returns 404 error when tag does not exist', async () => {
      await request(app.getHttpServer())
        .get('/tags/details/999999')
        .expect(404);
    });

    it('validate and convert parameters', async () => {
      const httpServer = app.getHttpServer();
      const badRequests = [
        '/tags/details/0',
        '/tags/details/bite',
        '/tags/details/NaN',
      ];
      const goodRequests = [
        `/tags/details/${random.arrayElement(mockedTags).id}`,
      ];

      await testRequests(httpServer, badRequests, 400);
      await testRequests(httpServer, goodRequests, 200);
    });
  });

  describe('/tags/search', () => {
    it.todo('returns tags including search query');

    it.todo('returns tags including search query and category');

    it.todo('returns empty array if nothing found');

    it.todo('validate and convert parameters');
  });

  afterAll(async () => {
    await app.close();
  });
});
