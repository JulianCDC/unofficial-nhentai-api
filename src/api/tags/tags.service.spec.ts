import { CacheModule, NotFoundException } from '@nestjs/common';
import { getConnectionToken, MongooseModule } from '@nestjs/mongoose';
import { Test } from '@nestjs/testing';
import { random } from 'faker';
import { filter, orderBy, slice } from 'lodash';
import { Connection } from 'mongoose';
import { CATEGORIES } from '../../common/enum/tag-categories.enum';
import { Tag, TagSchema } from '../../common/schemas/tag.schema';
import { TagWithCategory } from '../../common/types/tag-with-category.entity';
import {
  closeMongoConnection,
  DatabaseTestingModule,
} from '../../testing/database-testing/database-testing.module';
import { SeederService } from '../../testing/database-testing/seeder/seeder.service';
import { TagsService } from './tags.service';

describe('CategoriesController', () => {
  let tagsService: TagsService;
  let databaseConnection: Connection;
  let seededTags: Tag[];
  let seededWith: TagWithCategory[];

  const getNumberOfPages = (category: CATEGORIES) => {
    const filteredTags = seededTags.filter(tag => tag.category === category);
    return Math.ceil(filteredTags.length / pageSize);
  };

  const pageSize = 120;

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      controllers: [],
      providers: [TagsService],
      imports: [
        CacheModule.register({ ttl: 1 }),
        DatabaseTestingModule,
        MongooseModule.forFeature([{ name: Tag.name, schema: TagSchema }]),
      ],
    }).compile();

    tagsService = moduleRef.get(TagsService);
    databaseConnection = moduleRef.get(getConnectionToken());

    const seedingService = moduleRef.get(SeederService);
    [seededTags, seededWith] = await seedingService.seedTags();
  });

  afterEach(async () => {
    await databaseConnection.close();
    await closeMongoConnection();
  });

  describe('getPageCount', () => {
    it('returns correct page count', async () => {
      const numberOfPages = getNumberOfPages(CATEGORIES.artists);

      expect(await tagsService.getPageCount(CATEGORIES.artists)).toBe(
        numberOfPages,
      );
    });
  });

  describe('getTagsByPopularity', () => {
    const processTags = (sliceStart: number, sliceEnd: number) => {
      const filteredTags = filter(seededWith, [
        'category',
        CATEGORIES.artists.toString(),
      ]);
      const sortedTags = orderBy(filteredTags, 'tagged', 'desc');
      const slicedTags = slice(sortedTags, sliceStart, sliceEnd);

      return slicedTags;
    };

    it('returns tags sorted by popularity', async () => {
      const result = processTags(0, pageSize);

      await expect(
        tagsService
          .getTagsByPopularity(CATEGORIES.artists, 1)
          .then(tags => JSON.stringify(tags)),
      ).resolves.toBe(JSON.stringify(result));
    });

    it('works with random page', async () => {
      const numberOfPages = getNumberOfPages(CATEGORIES.artists);
      const randomPage = random.number({ min: 1, max: numberOfPages });

      const result = processTags(
        (randomPage - 1) * pageSize,
        (randomPage - 1) * pageSize + pageSize,
      );

      await expect(
        tagsService
          .getTagsByPopularity(CATEGORIES.artists, randomPage)
          .then(tags => JSON.stringify(tags)),
      ).resolves.toBe(JSON.stringify(result));
    });

    it('returns empty array when page out of range', async () => {
      const numberOfPages = getNumberOfPages(CATEGORIES.artists);

      await expect(
        tagsService.getTagsByPopularity(CATEGORIES.artists, numberOfPages + 1),
      ).resolves.toEqual([]);
    });
  });

  describe('getTagsByLetter', () => {
    const processTags = (letter: string) => {
      const filteredTags = filter(
        seededWith,
        tag =>
          tag.category === CATEGORIES.artists.toString() &&
          tag.name.toLowerCase().startsWith(letter.toLowerCase()),
      );
      const sortedTags = orderBy(filteredTags, 'name', 'asc');

      return sortedTags;
    };

    it('returns all tags starting with requested letter', async () => {
      const result = processTags('F');

      await expect(
        tagsService
          .getTagsByLetter(CATEGORIES.artists, 'F')
          .then(tags => JSON.stringify(tags)),
      ).resolves.toBe(JSON.stringify(result));
    });

    it('is case insensitive', async () => {
      const result = processTags('F');

      await expect(
        tagsService
          .getTagsByLetter(CATEGORIES.artists, 'F')
          .then(tags => JSON.stringify(tags)),
      ).resolves.toBe(JSON.stringify(result));

      await expect(
        tagsService
          .getTagsByLetter(CATEGORIES.artists, 'f')
          .then(tags => JSON.stringify(tags)),
      ).resolves.toBe(JSON.stringify(result));
    });

    it('returns empty array when letter is invalid', async () => {
      const result = [];

      await expect(
        tagsService.getTagsByLetter(CATEGORIES.artists, '$'),
      ).resolves.toEqual(result);
    });
  });

  describe('getTags', () => {
    const processTags = (sliceStart: number, sliceEnd: number) => {
      const filteredTags = filter(seededWith, [
        'category',
        CATEGORIES.artists.toString(),
      ]);
      const sortedTags = orderBy(filteredTags, 'name', 'asc');
      const slicedTags = slice(sortedTags, sliceStart, sliceEnd);

      return slicedTags;
    };

    it('returns all tags sorted by name', async () => {
      const result = processTags(0, pageSize);

      await expect(
        tagsService
          .getTags(CATEGORIES.artists, 1)
          .then(tags => JSON.stringify(tags)),
      ).resolves.toBe(JSON.stringify(result));
    });

    it('works with random page', async () => {
      const numberOfPages = getNumberOfPages(CATEGORIES.artists);
      const randomPage = random.number({ min: 1, max: numberOfPages });

      const result = processTags(
        (randomPage - 1) * pageSize,
        (randomPage - 1) * pageSize + pageSize,
      );

      await expect(
        tagsService
          .getTags(CATEGORIES.artists, randomPage)
          .then(tags => JSON.stringify(tags)),
      ).resolves.toBe(JSON.stringify(result));
    });

    it('returns empty array when page out of range', async () => {
      const numberOfPages = getNumberOfPages(CATEGORIES.artists);

      await expect(
        tagsService.getTags(CATEGORIES.artists, numberOfPages + 1),
      ).resolves.toEqual([]);
    });
  });

  describe('search', () => {
    const processTags = (query: string, category?: CATEGORIES) => {
      const filteredTags = seededWith.filter(tag => {
        const nameRegexp = new RegExp(`^.*${query}.*$`);

        if (category === undefined) return nameRegexp.test(tag.name);

        return (
          nameRegexp.test(tag.name) && tag.category === category.toString()
        );
      });
      const orderedTags = filteredTags.sort((a, b) => {
        const aName = a.name;
        const bName = b.name;

        return aName.indexOf(query) - bName.indexOf(query);
      });
      const slicedTags = orderedTags.slice(0, 10);

      return slicedTags;
    };

    it.skip('returns all tags matching search query', async () => {
      const result = processTags('ab');

      await expect(tagsService.search('ab')).resolves.toStrictEqual(result);
    });

    it.todo('returns all tags matching search query and category');
  });

  describe('getTagById', () => {
    it('returns requested tag', async () => {
      const randomTag = random.arrayElement(seededWith);

      await expect(
        tagsService.getTagById(randomTag.id).then(tag => JSON.stringify(tag)),
      ).resolves.toBe(JSON.stringify(randomTag));
    });

    it('throw NotFoundException when tag does not exist', async () => {
      await expect(
        tagsService.getTagById(Number.POSITIVE_INFINITY),
      ).rejects.toThrowError(NotFoundException);
    });
  });
});
