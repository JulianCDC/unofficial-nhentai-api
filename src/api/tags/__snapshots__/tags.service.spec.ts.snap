// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesController getPageCount returns correct page count 1`] = `3`;

exports[`CategoriesController getTagById returns requested tag 1`] = `
Object {
  "category": "characters",
  "id": 621918,
  "name": "voluptatem",
  "tagged": 40147,
  "uri": "https://emiliano.com",
}
`;

exports[`CategoriesController getTags returns all tags sorted by name 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 22330,
    "name": "ab",
    "tagged": 51935,
    "uri": "https://ozella.net",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 550947,
    "name": "aut",
    "tagged": 57974,
    "uri": "http://genoveva.name",
  },
  Object {
    "category": "artists",
    "id": 517593,
    "name": "aut",
    "tagged": 93487,
    "uri": "https://oswaldo.name",
  },
  Object {
    "category": "artists",
    "id": 128124,
    "name": "esse",
    "tagged": 99718,
    "uri": "https://torey.com",
  },
  Object {
    "category": "artists",
    "id": 39054,
    "name": "et",
    "tagged": 89460,
    "uri": "https://arvel.net",
  },
  Object {
    "category": "artists",
    "id": 505463,
    "name": "et",
    "tagged": 33328,
    "uri": "https://adela.biz",
  },
  Object {
    "category": "artists",
    "id": 144691,
    "name": "et",
    "tagged": 51036,
    "uri": "http://yadira.net",
  },
  Object {
    "category": "artists",
    "id": 593480,
    "name": "hic",
    "tagged": 59911,
    "uri": "http://adrienne.org",
  },
  Object {
    "category": "artists",
    "id": 878116,
    "name": "ipsam",
    "tagged": 68521,
    "uri": "https://connor.info",
  },
  Object {
    "category": "artists",
    "id": 916860,
    "name": "magni",
    "tagged": 52479,
    "uri": "http://aron.net",
  },
  Object {
    "category": "artists",
    "id": 538830,
    "name": "molestias",
    "tagged": 24621,
    "uri": "http://ryley.biz",
  },
  Object {
    "category": "artists",
    "id": 233622,
    "name": "omnis",
    "tagged": 96766,
    "uri": "https://dudley.com",
  },
  Object {
    "category": "artists",
    "id": 239847,
    "name": "quis",
    "tagged": 97001,
    "uri": "http://rosella.net",
  },
  Object {
    "category": "artists",
    "id": 186260,
    "name": "repellat",
    "tagged": 14675,
    "uri": "http://audrey.info",
  },
  Object {
    "category": "artists",
    "id": 561029,
    "name": "sunt",
    "tagged": 58575,
    "uri": "https://wade.info",
  },
  Object {
    "category": "artists",
    "id": 691876,
    "name": "tempore",
    "tagged": 95788,
    "uri": "http://kaci.biz",
  },
  Object {
    "category": "artists",
    "id": 322680,
    "name": "temporibus",
    "tagged": 69805,
    "uri": "https://sam.com",
  },
  Object {
    "category": "artists",
    "id": 927507,
    "name": "vel",
    "tagged": 39767,
    "uri": "https://cassandra.name",
  },
  Object {
    "category": "artists",
    "id": 573678,
    "name": "vel",
    "tagged": 23702,
    "uri": "https://stewart.net",
  },
]
`;

exports[`CategoriesController getTags works with any page 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 878116,
    "name": "ipsam",
    "tagged": 68521,
    "uri": "https://connor.info",
  },
  Object {
    "category": "artists",
    "id": 916860,
    "name": "magni",
    "tagged": 52479,
    "uri": "http://aron.net",
  },
  Object {
    "category": "artists",
    "id": 538830,
    "name": "molestias",
    "tagged": 24621,
    "uri": "http://ryley.biz",
  },
  Object {
    "category": "artists",
    "id": 233622,
    "name": "omnis",
    "tagged": 96766,
    "uri": "https://dudley.com",
  },
  Object {
    "category": "artists",
    "id": 239847,
    "name": "quis",
    "tagged": 97001,
    "uri": "http://rosella.net",
  },
  Object {
    "category": "artists",
    "id": 186260,
    "name": "repellat",
    "tagged": 14675,
    "uri": "http://audrey.info",
  },
  Object {
    "category": "artists",
    "id": 561029,
    "name": "sunt",
    "tagged": 58575,
    "uri": "https://wade.info",
  },
  Object {
    "category": "artists",
    "id": 691876,
    "name": "tempore",
    "tagged": 95788,
    "uri": "http://kaci.biz",
  },
  Object {
    "category": "artists",
    "id": 322680,
    "name": "temporibus",
    "tagged": 69805,
    "uri": "https://sam.com",
  },
  Object {
    "category": "artists",
    "id": 927507,
    "name": "vel",
    "tagged": 39767,
    "uri": "https://cassandra.name",
  },
  Object {
    "category": "artists",
    "id": 573678,
    "name": "vel",
    "tagged": 23702,
    "uri": "https://stewart.net",
  },
]
`;

exports[`CategoriesController getTagsByLetter is case insensitive 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 22330,
    "name": "ab",
    "tagged": 51935,
    "uri": "https://ozella.net",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 550947,
    "name": "aut",
    "tagged": 57974,
    "uri": "http://genoveva.name",
  },
  Object {
    "category": "artists",
    "id": 517593,
    "name": "aut",
    "tagged": 93487,
    "uri": "https://oswaldo.name",
  },
]
`;

exports[`CategoriesController getTagsByLetter is case insensitive 2`] = `
Array [
  Object {
    "category": "artists",
    "id": 22330,
    "name": "ab",
    "tagged": 51935,
    "uri": "https://ozella.net",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 550947,
    "name": "aut",
    "tagged": 57974,
    "uri": "http://genoveva.name",
  },
  Object {
    "category": "artists",
    "id": 517593,
    "name": "aut",
    "tagged": 93487,
    "uri": "https://oswaldo.name",
  },
]
`;

exports[`CategoriesController getTagsByLetter returns all tags starting with requested letter 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 22330,
    "name": "ab",
    "tagged": 51935,
    "uri": "https://ozella.net",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 550947,
    "name": "aut",
    "tagged": 57974,
    "uri": "http://genoveva.name",
  },
  Object {
    "category": "artists",
    "id": 517593,
    "name": "aut",
    "tagged": 93487,
    "uri": "https://oswaldo.name",
  },
]
`;

exports[`CategoriesController getTagsByPopularity returns tags sorted by popularity 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 128124,
    "name": "esse",
    "tagged": 99718,
    "uri": "https://torey.com",
  },
  Object {
    "category": "artists",
    "id": 239847,
    "name": "quis",
    "tagged": 97001,
    "uri": "http://rosella.net",
  },
  Object {
    "category": "artists",
    "id": 233622,
    "name": "omnis",
    "tagged": 96766,
    "uri": "https://dudley.com",
  },
  Object {
    "category": "artists",
    "id": 691876,
    "name": "tempore",
    "tagged": 95788,
    "uri": "http://kaci.biz",
  },
  Object {
    "category": "artists",
    "id": 517593,
    "name": "aut",
    "tagged": 93487,
    "uri": "https://oswaldo.name",
  },
  Object {
    "category": "artists",
    "id": 39054,
    "name": "et",
    "tagged": 89460,
    "uri": "https://arvel.net",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 322680,
    "name": "temporibus",
    "tagged": 69805,
    "uri": "https://sam.com",
  },
  Object {
    "category": "artists",
    "id": 878116,
    "name": "ipsam",
    "tagged": 68521,
    "uri": "https://connor.info",
  },
  Object {
    "category": "artists",
    "id": 593480,
    "name": "hic",
    "tagged": 59911,
    "uri": "http://adrienne.org",
  },
  Object {
    "category": "artists",
    "id": 561029,
    "name": "sunt",
    "tagged": 58575,
    "uri": "https://wade.info",
  },
  Object {
    "category": "artists",
    "id": 550947,
    "name": "aut",
    "tagged": 57974,
    "uri": "http://genoveva.name",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
  Object {
    "category": "artists",
    "id": 916860,
    "name": "magni",
    "tagged": 52479,
    "uri": "http://aron.net",
  },
  Object {
    "category": "artists",
    "id": 22330,
    "name": "ab",
    "tagged": 51935,
    "uri": "https://ozella.net",
  },
  Object {
    "category": "artists",
    "id": 144691,
    "name": "et",
    "tagged": 51036,
    "uri": "http://yadira.net",
  },
  Object {
    "category": "artists",
    "id": 927507,
    "name": "vel",
    "tagged": 39767,
    "uri": "https://cassandra.name",
  },
  Object {
    "category": "artists",
    "id": 505463,
    "name": "et",
    "tagged": 33328,
    "uri": "https://adela.biz",
  },
  Object {
    "category": "artists",
    "id": 538830,
    "name": "molestias",
    "tagged": 24621,
    "uri": "http://ryley.biz",
  },
  Object {
    "category": "artists",
    "id": 573678,
    "name": "vel",
    "tagged": 23702,
    "uri": "https://stewart.net",
  },
  Object {
    "category": "artists",
    "id": 186260,
    "name": "repellat",
    "tagged": 14675,
    "uri": "http://audrey.info",
  },
]
`;

exports[`CategoriesController getTagsByPopularity works with other pages 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 561029,
    "name": "sunt",
    "tagged": 58575,
    "uri": "https://wade.info",
  },
  Object {
    "category": "artists",
    "id": 550947,
    "name": "aut",
    "tagged": 57974,
    "uri": "http://genoveva.name",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
  Object {
    "category": "artists",
    "id": 916860,
    "name": "magni",
    "tagged": 52479,
    "uri": "http://aron.net",
  },
  Object {
    "category": "artists",
    "id": 22330,
    "name": "ab",
    "tagged": 51935,
    "uri": "https://ozella.net",
  },
  Object {
    "category": "artists",
    "id": 144691,
    "name": "et",
    "tagged": 51036,
    "uri": "http://yadira.net",
  },
  Object {
    "category": "artists",
    "id": 927507,
    "name": "vel",
    "tagged": 39767,
    "uri": "https://cassandra.name",
  },
  Object {
    "category": "artists",
    "id": 505463,
    "name": "et",
    "tagged": 33328,
    "uri": "https://adela.biz",
  },
  Object {
    "category": "artists",
    "id": 538830,
    "name": "molestias",
    "tagged": 24621,
    "uri": "http://ryley.biz",
  },
  Object {
    "category": "artists",
    "id": 573678,
    "name": "vel",
    "tagged": 23702,
    "uri": "https://stewart.net",
  },
  Object {
    "category": "artists",
    "id": 186260,
    "name": "repellat",
    "tagged": 14675,
    "uri": "http://audrey.info",
  },
]
`;

exports[`CategoriesController search returns all tags matching search query 1`] = `
Array [
  Object {
    "category": "characters",
    "id": 694399,
    "name": "incidunt",
    "tagged": 10233,
    "uri": "https://harrison.org",
  },
  Object {
    "category": "artists",
    "id": 878116,
    "name": "ipsam",
    "tagged": 68521,
    "uri": "https://connor.info",
  },
  Object {
    "category": "tags",
    "id": 412538,
    "name": "ipsam",
    "tagged": 14212,
    "uri": "https://otis.org",
  },
  Object {
    "category": "characters",
    "id": 593064,
    "name": "ipsum",
    "tagged": 5924,
    "uri": "https://marco.com",
  },
  Object {
    "category": "characters",
    "id": 941836,
    "name": "hic",
    "tagged": 6406,
    "uri": "http://fiona.name",
  },
  Object {
    "category": "characters",
    "id": 73364,
    "name": "minima",
    "tagged": 44613,
    "uri": "https://damaris.com",
  },
  Object {
    "category": "groups",
    "id": 881340,
    "name": "enim",
    "tagged": 35653,
    "uri": "https://nicholas.info",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 916860,
    "name": "magni",
    "tagged": 52479,
    "uri": "http://aron.net",
  },
  Object {
    "category": "characters",
    "id": 457204,
    "name": "accusantium",
    "tagged": 53441,
    "uri": "https://tatyana.info",
  },
]
`;

exports[`CategoriesController search returns all tags matching search query and category 1`] = `
Array [
  Object {
    "category": "artists",
    "id": 878116,
    "name": "ipsam",
    "tagged": 68521,
    "uri": "https://connor.info",
  },
  Object {
    "category": "artists",
    "id": 593480,
    "name": "hic",
    "tagged": 59911,
    "uri": "http://adrienne.org",
  },
  Object {
    "category": "artists",
    "id": 239847,
    "name": "quis",
    "tagged": 97001,
    "uri": "http://rosella.net",
  },
  Object {
    "category": "artists",
    "id": 233622,
    "name": "omnis",
    "tagged": 96766,
    "uri": "https://dudley.com",
  },
  Object {
    "category": "artists",
    "id": 748164,
    "name": "aperiam",
    "tagged": 75014,
    "uri": "https://yasmin.name",
  },
  Object {
    "category": "artists",
    "id": 916860,
    "name": "magni",
    "tagged": 52479,
    "uri": "http://aron.net",
  },
  Object {
    "category": "artists",
    "id": 538830,
    "name": "molestias",
    "tagged": 24621,
    "uri": "http://ryley.biz",
  },
  Object {
    "category": "artists",
    "id": 322680,
    "name": "temporibus",
    "tagged": 69805,
    "uri": "https://sam.com",
  },
  Object {
    "category": "artists",
    "id": 918732,
    "name": "accusantium",
    "tagged": 54954,
    "uri": "http://tierra.net",
  },
]
`;
